Allow variable-specific transforms

ricardoV94 opened the pull request:
Closes #32

I implemented an opt container `TransformValuesOpt` that allows one to specify what transforms to apply to different value variables. This example illustrates the interface:

```python
import aesara.tensor as at
from aeppl import joint_logprob
from aeppl.transforms import DEFAULT_TRANSFORM, LogOddsTransform, TransformValuesOpt

x_rv = at.random.uniform(0, 1, name="x")
y_rv = at.random.uniform(x_rv, 5, name="y")
z_rv = at.random.uniform(y_rv, 10, name="z")

x = x_rv.clone()
y = y_rv.clone()
z = z_rv.clone()

transform_opt = TransformValuesOpt(
    {
        x: LogOddsTransform(),
        y: DEFAULT_TRANSFORM,
        z: None,
    }
)

joint_logprob(
    z_rv,
    {x_rv: x, y_rv: y, z_rv: z},
    extra_rewrites=transform_opt
)
```

ricardoV94 commented:
Apologies for the large shuffling around and renaming of objects. The file was getting quite hard navigate. Hopefully it is now a bit more readable.

ricardoV94 commented:
Okay, I split the PR into two commits. The first commit is only a refactoring of `transforms.py` and does not change any functionality. The second commit adds the new functionality.

brandonwillard commented on aeppl/transforms.py:
```diff
@@ -47,50 +51,57 @@ def class_transformed_rv(op, node):
+    return new_op_type
+
+
+# =============
+# Transform Opt
+# =============
```
Comment-based sectional delimiters are just a personal style choice, and not something that particularly helps the readability/navigation of the code, so let's not add these sorts of things.

ricardoV94 replied:
I'll remove them

brandonwillard commented on aeppl/transforms.py:
```diff
@@ -56,25 +54,32 @@ class TransformedRV(RandomVariable, metaclass=TransformedRVMeta):
 
 @singledispatch
-def _default_transformed_rv(
+def _get_default_transformed_rv_op(
```
This looks like a purely refactoring change.  Is it necessary for the functionality changes implied by this commit?  If so, you'll need to clarify; otherwise, it would be better in a separate commit, because then we can focus on the functionality itself.

ricardoV94 replied:
The returned object of that method is no longer a node but the `TransformedRV` `Op`. The reason for that is that I have to create the node for non default transforms anyway in the local optimizer, so there was no reason to specialize the dispatch method as much

brandonwillard replied:
We definitely don't want to start basing our object names so directly on their exact input and return types.  That information is suffiently conveyed by the type information and docstrings; the names really only need to reflect the basic actions/operations, and some ambiguity/inaccuracy is fine as long as it isn't outright contradictory.

ricardoV94 replied:
I am not usually this nitpick,l with the names.

It is just that in this module there are so many things called `transform*` that it becomes hard to keep track. 

In any case, I am fine with reverting this one. I don't like the long name either.

ricardoV94 replied:
On the pro side, it aligns with the `create_transformed_rv_op` name, and those are returning the same type of object now

brandonwillard commented on aeppl/transforms.py:
```diff
@@ -285,67 +365,82 @@ def log_jac_det(self, value, *inputs):
 # ======================
 
 
-TransformedUniformRV = create_default_transformed_rv_op(
+TransformedUniformRV = create_transformed_rv_op(
```
These renamings appear to be undoing renamings from an earlier commit of yours: 52dd503.  Is this intentional, or are these changes unintended regressions that snuck into the commit?

ricardoV94 replied:
Intentional, the previous change made the name better reflect was the method was doing at the time of the previous commit, and the same applies here.

ricardoV94 replied:
They are not an exact revert. It has the new `op` suffix

brandonwillard reviewed (dismissed):
There are a lot of style and refactoring changes that don't appear to be necessary and distract from the functional changes (both within and across commits), but the main idea/approach (i.e. `TransformValuesMapping`) seems fine.

I don't know if we need the `GlobalOptimizer`-like `TransformValuesOpt` class to wrap the use of `TransformValuesMapping`, though.

If you put the unrelated refactoring and style changes into separate commits, we can review and consider those, but that's going to slow the acceptance of the important functional changes that define _this_ PR, so it would be much better to put those in a separate PR that we can review independently.

brandonwillard commented on aeppl/transforms.py:
```diff
@@ -66,76 +34,231 @@ def log_jac_det(self, value: TensorVariable, *inputs) -> TensorVariable:
+
+
+class TransformValuesMapping(Feature):
+    r"""Store dictionary indicating the transforms to apply to each value variable"""
```
```suggestion
    r"""A `Feature` that maintains a map between value variables and their transforms."""
```

brandonwillard commented on aeppl/transforms.py:
```diff
@@ -66,76 +34,231 @@ def log_jac_det(self, value: TensorVariable, *inputs) -> TensorVariable:
+    r"""Store dictionary indicating the transforms to apply to each value variable"""
+
+    def __init__(self, transform_values):
+        self.transform_values = transform_values
```
It would make more sense to call this variable something like `values_to_transforms`, or `vars_to_transforms`.  The current name implies somethings else.

ricardoV94 replied:
I like `values_to_transforms`

ricardoV94 commented:
> I don't know if we need the `GlobalOptimizer`-like `TransformValuesOpt` class to wrap the use of `TransformValuesMapping`, though.

My initial approach involved passing the feature and the rewrite as a dictionary but it made the code a bit more messy in the `joint_logprob`.

Another option would be to have an `extra_ features` like `extra_rewrites` which is called before, but I am not sure that looks better...

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -47,50 +51,52 @@ def class_transformed_rv(op, node):
-        # return at.log(at.abs_(jac))
-        phi_inv = self.backward(value, *inputs)
-        return at.log(at.nlinalg.det(at.atleast_2d(jacobian(phi_inv, [value]))))
+@singledispatch
```
This was moved here, so that in the next commit I can change the return type to `Optional[TransformedRV]` without additional changes

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -56,25 +54,32 @@ class TransformedRV(RandomVariable, metaclass=TransformedRVMeta):
 
 @singledispatch
-def _default_transformed_rv(
+def _default_transformed_rv_op(
```
I still think it makes sense to add the `_op` suffix in this commit, as it exists in the related `create_transformed_rv_op`. 

I imagine the original name distinction was exactly because of the different returned types, and that if we were implementing this from scratch we would have chosen similar signatures...

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -66,76 +34,218 @@ def log_jac_det(self, value: TensorVariable, *inputs) -> TensorVariable:
+    def on_attach(self, fgraph):
+        if hasattr(fgraph, "transform_values_mapping"):
+            raise ValueError(
+                f"{fgraph} already has the `TransformValuesMapping` feature attached."
```
Is it worth it to add a test for this `ValueError`? There is something similar in the `PreserveRVMappings` that is also not being tested AFAIK. 

Also, on the Aesara docs, it suggests we should raise the more specific `AlreadyThere` exception: https://aesara.readthedocs.io/en/latest/library/graph/fgraph.html?highlight=feature#aesara.graph.features.Feature.on_attach

Should we do that?

brandonwillard replied:
Yeah, the `AlreadyThere` approach sounds better (assuming it really was/is fully supported).

brandonwillard commented:
I've simplified this commit and added a missing test for graphs containing multi-output `Op`s that do not also have default outputs (e.g. `SVD`).

Right now, my only real concern with this implementation is that it requires one to specify that a default transform should be used.  Instead, it should probably apply the default transforms without the need to specify them.

rlouf commented:
On this last point: should the decision of what constitutes a default transform and whether to apply it be made at the level of this library? It feels closer to the application than what I thought `aeppl` was going to be.

My inclination would be to apply no transformation if nothing is specified.

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -135,6 +158,63 @@ def default_transform(fgraph: FunctionGraph, node: Node) -> Optional[List[Node]]
+
+        """
+
+        self.values_to_transforms = values_to_transforms.copy()
```
I was copying the dict when passing it to the feature so that the optimizer could be reused without having to reinitialize it (otherwise the class copy of the dict is spent inplace).

I feel that is less error prone (eg if in an iterative environment like Jupyter one makes the opt in a separate cell and runs the joint logprob twice the second one would silently not do anything

ricardoV94 replied:
A test should be added if we want to support this functionality

brandonwillard replied:
The only logical reason for this step was the check being performed in the original code and its accompanying error, and, since the latter was already questionable (and now just a warning), the `list.pop` and `copy` steps should be removed entirely.

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -89,6 +91,13 @@ def transformed_logprob(op, value, *inputs, name=None, **kwargs):
+
+
 @local_optimizer(tracks=None)
 def default_transform(fgraph: FunctionGraph, node: Node) -> Optional[List[Node]]:
```
Calling this opt `default_transform` feels wrong.

ricardoV94 replied:
Also don't you prefer the more specific return type hint?

ricardoV94 replied:
That is `Optional[List[TransformedRV]]`

brandonwillard replied:
These are unrelated changes, but we should create some PRs for them.

ricardoV94 replied:
Done: https://github.com/aesara-devs/aeppl/issues/43

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -135,6 +158,63 @@ def default_transform(fgraph: FunctionGraph, node: Node) -> Optional[List[Node]]
+
+        Raises
+        ======
+        RuntimeError
```
It's just a warning now

brandonwillard replied:
Yes, by design; it's not clear that an error is more suitable.

ricardoV94 replied:
I meant that the docstrings are incorrect now

ricardoV94 commented:
> Right now, my only real concern with this implementation is that it requires one to specify that a default transform should be used.  Instead, it should probably apply the default transforms without the need to specify them.

I'd rather require users to be explicit. Transforms already have a tendency to be very invisible / surprising. For other libraries like PyMC it won't matter, as the dict will be created programatically.

We could add a less flexible opt that Only applies default transforms (just inherit from this and use the value_vars in the fgraph)

ricardoV94 commented:
> On this last point: should the decision of what constitutes a default transform and whether to apply it be made at the level of this library? It feels closer to the application than what I thought `aeppl` was going to be.

The defaults can be overwritten easily by another user / library. Otherwise I think they are pretty standard defaults


> My inclination would be to apply no transformation if nothing is specified.

Me too

brandonwillard commented:
> On this last point: should the decision of what constitutes a default transform and whether to apply it be made at the level of this library? It feels closer to the application than what I thought `aeppl` was going to be.
> 
> My inclination would be to apply no transformation if nothing is specified.

When we say "default" we mean "the default for the transformation rewrites feature".  In AePPL, transformations aren't applied unless the user explicitly adds the transformation rewrites (see the unit tests for examples).

Otherwise, regarding the default transforms performed by the transform rewriter: those are configurable by customizing a generic/single dispatch function.

brandonwillard commented:
> I'd rather require users to be explicit. Transforms already have a tendency to be very invisible / surprising. For other libraries like PyMC it won't matter, as the dict will be created programatically.

OK, we'll go with that.

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -240,8 +320,24 @@ def log_jac_det(self, value, *inputs):
     rv_op: Op,
     transform: RVTransform,
+    *,
+    default: bool = True,
```
I feel the default should be there option with least side effects (and these dispatch side effects are quite significant imho). Even though it definitely makes things more verbose for what is the most common use.

ricardoV94 commented:
This seems ready, only thing left to do is deciding on whether the `default` of `create_default_transformed_rv_op` should be `True` or `False`.

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -89,52 +90,113 @@ def transformed_logprob(op, value, *inputs, name=None, **kwargs):
 
-    # Get the old value variable and remove it from our value variables map
-    rv_var = node.outputs[1]
-    old_value_var = rv_map_feature.rv_values.pop(rv_var)
```
There was no need to pop the `old_value_var` since we already have it from above and we are replacing it with the `new_value_var` at the end anyway.

brandonwillard replied:
If we don't pop them, then we'll end up with a map that has entries for variables that are no longer in the graph, and&mdash;unless I'm missing something&mdash;we don't want/need those.

ricardoV94 replied:
The original rewrite was reintroducing it in the dictionary, but associated with the new value. See the other comment.

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -237,11 +299,27 @@ def log_jac_det(self, value, *inputs):
 
 
-def create_default_transformed_rv_op(
+def _create_transformed_rv_op(
```
This method is going to be further expanded in #26 so it seems reasonable to make it private and just expose the user facing `create_default_transformed_rv_op` with the fixed default kwargs (via the partial)

brandonwillard replied:
This change doesn't make the function private in any functional sense; it's just a naming convention.  If we're concerned about the things we're exposing to users, the best we can do is set `__all__`.

Regardless, this is a minor style-based refactoring that should be in a separate PR/commit.

brandonwillard reviewed (dismissed):
It looks like you (re)added some refactoring changes to this commit.

brandonwillard commented:
I put the most recent additions into a separate commit.

brandonwillard commented on aeppl/transforms.py:
```diff
@@ -89,52 +91,115 @@ def transformed_logprob(op, value, *inputs, name=None, **kwargs):
-        new_value_var.name = f"{old_value_var.name}_{transform.name}"
-
     rv_map_feature.rv_values[new_rv_var] = new_value_var
+    rv_map_feature.rv_values[rv_var] = new_value_var
```
Do we need this association?

ricardoV94 replied:
It was there originally. I am not sure we do though

ricardoV94 commented on aeppl/transforms.py:
```diff
@@ -89,52 +91,115 @@ def transformed_logprob(op, value, *inputs, name=None, **kwargs):
     # This means that we need to replace all instance of the old value variable
     # with "inversely/un-" transformed versions of itself.
-    new_value_var = transform.backward(old_value_var, *trans_node.inputs)
-    rv_map_feature.rv_values[rv_var] = new_value_var
```
This was the line where it was being reintroduced to the mapping feature

ricardoV94 commented:
I removed the old RV from the mapping and cleaned some changes that were creeping into the second commit. Let me know it it looks good @brandonwillard
